MediaRecorder.bs

```diff
@@ -241,6 +249,33 @@ interface MediaRecorder : EventTarget {
     the value might be surpassed, not achieved, or only be achieved over a long
     period of time.</li>
 
+    <li>Let |videoKeyFrameIntervalDuration| be
+    |recorder|.[=[[VideoKeyFrameIntervalDuration]]=], and let
+    |videoKeyFrameIntervalCount| be
+    |recorder|.[=[[VideoKeyFrameIntervalCount]]=]. The UA SHOULD constrain the
+    configuration of |recorder| so that the video encoder follows the below rules:
+
+    <ul>
+    <li>If |videoKeyFrameIntervalDuration| is not <code>null</code> and |videoKeyFrameIntervalCount| is <code>null</code>,
+    the video encoder produces a keyframe on the first frame arriving after |videoKeyFrameIntervalDuration|
+    seconds elapsed since the last key frame.</li>
+
+    <li>If |videoKeyFrameIntervalCount| is not <code>null</code> and |videoKeyFrameIntervalDuration| is <code>null</code>,
+    the video encoder produces a keyframe on the first frame arriving after |videoKeyFrameIntervalCount|
+    frames passed since the last key frame.</li>
+
+    <li>If |videoKeyFrameIntervalDuration| and |videoKeyFrameIntervalCount| are both not <code>null</code>,
+    the video encoder produces a keyframe on the first frame arriving after
+    both |videoKeyFrameIntervalDuration| seconds elapsed AND at least |videoKeyFrameIntervalCount|
+    frames passed since the last key frame.</li>
```

Comment:
Thanks, this seems clear!

I guess my remaining question is a content one, for folks like @Pehrsons: is this "both" behavior is usable? The alternative would be to throw if both are provided (one less thing to test). Maybe this was already covered in discussions and I missed it.

In any case I'm not gonna block on it.

Reply:
Thanks! All done now I hope. 

The rationale was discussed with Harald before on this PR:

"The rationale is to support variable-rate sources where you get low per-frame quality (QP) in stretches where FPS is low with a pure time-based control, since the ratio of keyframes (in which quality is lower) to delta frames (in which quality is enhanced) would be higher in the low FPS stretches. See also earlier comment in https://github.com/w3c/mediacapture-record/pull/216#issuecomment-1415989756"

Reply:
I don't have much thought on using both. I guess mainly because I don't really see a use case for a time-based keyframe interval. I've only seen mentioned the use case in [this comment](https://github.com/w3c/mediacapture-record/pull/216#issuecomment-1415989756) and think that with a framecount-based interval one could achieve similar results by having identical (or [negative](https://www.ietf.org/archive/id/draft-ietf-cellar-matroska-16.html#section-11.2-8)) presentation timestamps for the necessary frames before the desired cut without too much storage overhead, since the overhead would only consist of non-keyframes.

I guess a time-based interval could still make sense if there is precedence in the industry, which there seems to be.
As far as a use-case for setting both framecount- and time-based intervals goes, not sure.

@youennf [might](https://github.com/w3c/mediacapture-record/pull/216#issuecomment-1441955962) have more thoughts.

Reply:
I would tend to restrict to either one of the two options for now.
This restriction does not seem to impact the use cases we care.
It will be easy to remove this restriction later on if we see value in this.

Reply:
@Pehrsons 

> IIRC Firefox sets a keyframe interval of 10 seconds, for historical reasons [1], and occasionally recalculates the frame-interval based on recent incoming framerate, so it can configure the encoder with this.

> I guess mainly because I don't really see a use case for a time-based keyframe interval.

These statements seem a bit contradictory (time constraints imposed by the muxer implementation are mentioned. When the user has such time constraints to avoid transcoding the stream, a sole frame based control would force the need for framerate estimation or assumptions in JS (is this always possible?).

The "both" case was introduced in this discussion as a simple quality compromise solution for dealing with variable framerate. Since I have no live request for this feature right now I'm happy to punt that for later. I uploaded a new version which ignore either if both are set.

Reply:
In Firefox I intend to (eventually) allow the muxer to write a new cluster without a keyframe, thereby removing the need for the encoder to use a time-based key frame interval. I hope this resolves the contradiction.